extensions-core/namespace-lookup/src/main/java/io/druid/query/extraction/NamespaceLookupExtractorFactory.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.druid.query.extraction;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.google.api.client.repackaged.com.google.common.base.Preconditions;
+import com.google.common.base.Function;
+import com.google.inject.name.Named;
+import com.metamx.common.logger.Logger;
+import io.druid.query.extraction.namespace.ExtractionNamespace;
+import io.druid.query.lookup.LookupExtractor;
+import io.druid.query.lookup.LookupExtractorFactory;
+import io.druid.server.namespace.NamespacedExtractionModule;
+import io.druid.server.namespace.cache.NamespaceExtractionCacheManager;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+@JsonTypeName("namespace")
+public class NamespaceLookupExtractorFactory implements LookupExtractorFactory
+{
+  private static final Logger log = new Logger(NamespaceLookupExtractorFactory.class);
+
+  final ExtractionNamespace extractionNamespace;
+  final NamespaceExtractionCacheManager manager;
+  LookupExtractor lookupExtractor;
+
+  @JsonCreator
+  public NamespaceLookupExtractorFactory(
+      @JsonProperty("extractionNamespace")ExtractionNamespace extractionNamespace,
+      @JacksonInject NamespacedExtractionModule.NamespaceStaticConfiguration namespaceStaticConfiguration,
+      @JacksonInject @Named("dimExtractionNamespace")
+      Function<String, Function<String, String>> fnMaker,
+      @JacksonInject @Named("dimReverseExtractionNamespace")
+      Function<String, Function<String, List<String>>> reverseFnMaker
+      )
+  {
+    this.extractionNamespace = Preconditions.checkNotNull(extractionNamespace,
+        "extractionNamespace should be specified");
+    this.manager = namespaceStaticConfiguration.getManager();
+    this.lookupExtractor = new NamespacedExtractor(fnMaker, reverseFnMaker, extractionNamespace.getNamespace());
+  }
+
+  @Override
+  public boolean start()
```

Comment:
I feel like what should be happening here is that this object creates the schedule, and an "update" is performed by the thing managing LookupExtractorFactories,

As such, this method could call a `manager.schedule(String id, ExtractionNamespace extractionNamespace, long waitForFirstRun)` kind of method that returns after the extractionNamespace has been populated in the cache at least once, or until waitForFirstRun has timed out waiting.

The id passed does not need to be the ID used in a query, it can simply be a unique ID this NamespaceExtractionFactory uses to reference its part of the cache.

As such this class should have a field which keeps track of its ID used in the cache manager.

On close the delete call to the manager should delete the ID cache.